src/Lecture4.hs

```diff
@@ -134,8 +139,42 @@ errors. We will simply return an optional result here.
 🕯 HINT: Use the 'readMaybe' function from the 'Text.Read' module.
 -}
 
+-- Actually was a bit lazy and just googled most easist solution to understand 🤷‍♂️😅
+split :: Char -> String -> [String]
+split _ [] = [""]
+split c (x:xs)
+    | x == c = "" : rest
+    | otherwise = (x : head rest) : tail rest
+    where
+        rest = split c xs
+
+dropSpaces :: [Char] -> [Char] 
+dropSpaces = filter (not . isSpace)
+
 parseRow :: String -> Maybe Row
-parseRow = error "TODO"
+parseRow str = do 
+      let strings = split ',' str
+      when (length strings > 3) Nothing
+      name <- safeIndex 0 strings
+      when (null name) Nothing
+      t <- checkType (safeIndex 1 strings)
+      c <- checkInt (safeIndex 2 strings)
+      when (c < 0) Nothing
+      return (Row name t c)
+   where 
+      -- Don't know how to make polymorphic for now...
+      checkType :: Maybe String -> Maybe TradeType
+      checkType string = case string of
+         Nothing -> Nothing
+         Just s -> readMaybe (dropSpaces s) 
+      checkInt :: Maybe String -> Maybe Int
+      checkInt string = case string of
+         Nothing -> Nothing
+         Just s -> readMaybe (dropSpaces s)
+      safeIndex :: Int -> [String] -> Maybe String
+      safeIndex i xs
+        | (i > -1) && (length xs > i) = Just (xs!!i)
+        | otherwise = Nothing
```

Comment:
You have an excellent monadic implementation and even with the `do`-notation for `Maybe` 🔝 🏆  

I haven't explained the following during lectures but there's a nice way to write this parsing syntax using `do`-notation, `Monad` instance for `Maybe` and the `MonadFail` desugaring trick 🧙🏻 

You can read more about this trick here:

* https://kowainik.github.io/posts/haskell-mini-patterns#monadfail-sugar

Applying this programming pattern and a few helper functions, the `parseRow` function will look like this:

```suggestion
parseRow str = do
    [name, rawTrade, rawCost] <- Just (split ',' str)
    guard $ name /= ""
    t <- readMaybe rawTrade
    c <- readMaybe rawCost
    guard $ c >= 0
    name (Row name t c)
```

Reply:
Oh, thx! Have put it to my reading list. Still so many things to learn (and that's good 🙂).